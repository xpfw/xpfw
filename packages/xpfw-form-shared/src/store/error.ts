import { IFieldError, IFormError } from "@xpfw/validate"
import { get, set } from "lodash"
import { autoSubscribe, AutoSubscribeStore, key, StoreBase } from "resub"
import { IPersistableStore } from "resub-persist"

@AutoSubscribeStore
export class FormErrorStore extends StoreBase implements IPersistableStore {
    public name = "formErrors"
    private errors: {[path: string]: IFormError | IFieldError | undefined} = {}

    public getPropKeys() { return ["errors"] }

    public setError(valuePath: string, value: any) {
      set(this.errors, valuePath, value)
      this.trigger(valuePath)
    }

    @autoSubscribe
    public getError(@key path: string): any {
      return get(this.errors, path)
    }
}

export default new FormErrorStore()
