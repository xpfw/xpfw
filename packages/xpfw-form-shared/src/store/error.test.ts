import { matchStoreState } from "resub-persist"
import { FormErrorStore as FormErrorStoreCreator } from "./error"

const FormErrorStore = new FormErrorStoreCreator()
const errPath = "errpath"
const myErr = {my: "error"}

test("Cover Form Store", () => {
  matchStoreState(FormErrorStore, "Initial State")
  FormErrorStore.setError(errPath, myErr)
  matchStoreState(FormErrorStore, "Set Error")
  expect(FormErrorStore.getError(errPath)).toBe(myErr)
})
