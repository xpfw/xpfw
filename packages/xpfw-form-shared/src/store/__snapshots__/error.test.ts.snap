// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cover Form Store: Initial State 1`] = `
Object {
  "errors": Object {},
}
`;

exports[`Cover Form Store: Set Error 1`] = `
Object {
  "errors": Object {
    "errpath": Object {
      "my": "error",
    },
  },
}
`;
